runtime/interpreter/statictype.go

```diff
@@ -473,28 +473,153 @@ outer:
 	return t.Type.Equal(otherRestrictedType.Type)
 }
 
+// Authorization
+
+type Authorization interface {
+	isAuthorization()
+	String() string
+	MeteredString(common.MemoryGauge) string
+	Equal(auth Authorization) bool
+	Encode(e *cbor.StreamEncoder) error
+}
+
+type Unauthorized struct{}
+
+var UnauthorizedAccess Authorization = Unauthorized{}
+
+func (Unauthorized) isAuthorization() {}
+
+func (Unauthorized) String() string {
+	return ""
+}
+
+func (Unauthorized) MeteredString(memoryGauge common.MemoryGauge) string {
+	memoryGauge.MeterMemory(common.NewRawStringMemoryUsage(0))
+	return ""
+}
+
+func (Unauthorized) Equal(auth Authorization) bool {
+	_, ok := auth.(Unauthorized)
+	return ok
+}
+
+type EntitlementSetAuthorization struct {
+	Entitlements []common.TypeID
+	Kind         sema.EntitlementSetKind
+}
+
+var _ Authorization = EntitlementSetAuthorization{}
+
+func NewEntitlementSetAuthorization(memoryGauge common.MemoryGauge, entitlements []common.TypeID, kind sema.EntitlementSetKind) EntitlementSetAuthorization {
+	common.UseMemory(memoryGauge, common.MemoryUsage{
+		Kind:   common.MemoryKindEntitlementSetStaticAccess,
+		Amount: uint64(len(entitlements)),
+	})
+
+	return EntitlementSetAuthorization{Entitlements: entitlements, Kind: kind}
+}
+
+func (EntitlementSetAuthorization) isAuthorization() {}
+
+func (e EntitlementSetAuthorization) string(stringer func(common.TypeID) string) string {
+	var builder strings.Builder
+	builder.WriteString("auth(")
+	var separator string
+
+	if e.Kind == sema.Conjunction {
+		separator = ", "
+	} else if e.Kind == sema.Disjunction {
+		separator = " | "
+	}
+
+	for i, entitlement := range e.Entitlements {
+		builder.WriteString(stringer(entitlement))
+		if i < len(e.Entitlements) {
+			builder.WriteString(separator)
+		}
+	}
+	builder.WriteString(") ")
+	return builder.String()
+}
+
+func (e EntitlementSetAuthorization) String() string {
+	return e.string(func(ti common.TypeID) string { return string(ti) })
+}
+
+func (e EntitlementSetAuthorization) MeteredString(memoryGauge common.MemoryGauge) string {
+	memoryGauge.MeterMemory(common.AuthStringMemoryUsage)
+	return e.string(func(ti common.TypeID) string {
+		memoryGauge.MeterMemory(common.NewRawStringMemoryUsage(len(ti)))
+		return string(ti)
+	})
+}
+
+func (e EntitlementSetAuthorization) Equal(auth Authorization) bool {
+	if auth, ok := auth.(EntitlementSetAuthorization); ok {
+		for i, entitlement := range e.Entitlements {
+			if auth.Entitlements[i] != entitlement {
+				return false
+			}
+		}
+		return e.Kind == auth.Kind
+	}
+	return false
+}
+
+type EntitlementMapAuthorization struct {
+	TypeID common.TypeID
+}
+
+var _ Authorization = EntitlementMapAuthorization{}
+
+func NewEntitlementMapAuthorization(memoryGauge common.MemoryGauge, id common.TypeID) EntitlementMapAuthorization {
+	common.UseMemory(memoryGauge, common.NewConstantMemoryUsage(common.MemoryKindEntitlementMapStaticAccess))
```

Comment:
Could re-use this constant memory usage (`common.NewConstantMemoryUsage(common.MemoryKindEntitlementMapStaticAccess)`)
Have similar pre-defined ones in `common/metering.go`